src/preamble/01_what_do_we_mean_when_we_say_cabal.md

```diff
@@ -1 +1,96 @@
+different Haskell implementations. So it provides abstractions of features
+present in different Haskell implementations and wherever possible it is best to
+take advantage of these to increase portability. For example one of the pieces
+of information an author can put in the package’s .cabal file is what language
+extensions the code uses. This is preferable to specifying flags for a specific
+compiler as it allows Cabal to pick the right flags for the Haskell
+implementation that the user picks. It also allows Cabal to figure out if the
+language extension is even supported by the Haskell implementation that the user
+picks. Where compiler-specific options are needed, there is an “escape hatch”
+available. The developer can specify implementation-specific options and more
+generally there is a configuration mechanism to customise many aspects of how a
+package is built depending on the Haskell implementation, the operating system,
+computer architecture and user-specified configuration flags.
+
+There are various
+[Haskell implementations](https://wiki.haskell.org/Implementations), but we have
+now pretty much all converged on [GHC](https://www.haskell.org/ghc/) as the
+standard Haskell implementation.
+
+## The CABAL library
+
+Cabal stands for Common Architecture for Building Applications and Libraries.
+This is the library that provides functionality that allows the information in
+the .cabal files to be put to use. The Cabal library contains the code for
+parsing .cabal files and operations for building haskell packages. Cabal the
+library, by convention, is written with Capitalization case. Cabal can take a
+haskell dependency graph (of external and internal modules) and use GHC to build
+it.
+
+## The binary cabal-install (cli tool)
+
+The cabal binary or more accurately `cabal-install` is the command-line tool
+that provides a user interface for dealing with Haskell packages. cabal-install
+makes use of Cabal the library to do its job.
+
+cabal-install is a frontend to Cabal. It makes it possible to build Haskell
+projects whose sets of dependencies might conflict with each other within the
+confines of a single system. A package's .cabal file provides a constraint on
+the version of various libraries (version bounds) that they expect. When
+cabal-install is asked to build a project, by default it looks at the
+dependencies specified in its .cabal file and uses a dependency solver to figure
+out a set of packages and package versions that satisfy it. This set of packages
+is drawn from Hackage. The chosen versions of the dependenices will be installed
+and indexed in a database in the cabal directory (`~/.cabal`).
+
+Conflicts between dependencies are avoided by indexing the installed packages
+according to their version and other relevant configuration options. This allows
+different projects to retrieve the dependency versions they need.
+
+stack is also a command-line tool that depends on the Cabal Library, and hence
```

Comment:
Let's be high-minded and also compliment ourselves saying something like "stack is another mature command-line tool, with somewhat different goals and trade-offs, that depends...".